### PR TITLE
Warn when inlining a temporary into a conditional

### PR DESCRIPTION
**Pull request by YairHalberstadt**
I had a suggestion to inline a temporary into a Debug.Assert, which would have semantically changed the meaning of the program, since the expression would not have been called in Release.

The suggestion was to change:
```csharp
var x = stack.Pop();
Debug.Assert(x == this);
```
to
```csharp
Debug.Assert(stack.Pop() == this);
```

Inline temporary expression can often change semantics, but in this case I think it is worth warning about, since it's impossible to tell semantics have changed without inspecting the method being called. In the above case the two have identical semantics in Debug (assuming `stack` is a local).

Here is the warning:

image

I only warn when the expression is being inlined into a single location, with the assumption being that if an expression is being inlined into multiple locations, it presumably doesn't matter how often it is called, and is probably side effect free from the perspective of the caller.

This change required moving MethodSymbol.IsConditional to the public API - I don't know if this is a problem, since it can trigger binding.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv @333fred for compiler change.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs [hunk 1]
```diff
@@ -256,5 +256,13 @@ public interface IMethodSymbol : ISymbol
         /// anonymous delegate associated with it.
         /// </summary>
         INamedTypeSymbol AssociatedAnonymousDelegate { get; }
+
+        /// <summary>
+        /// Returns a flag indicating whether this symbol has at least one applied/inherited conditional attribute.
+        /// </summary>
+        /// <remarks>
+        /// May Force binding and decoding of attributes.
+        /// </remarks>
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs]: remove this remark.  i don't think it has any concept at our public layer.

File: src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs [hunk 2]
```diff
@@ -4746,5 +4746,54 @@ public async Task Selections3()
...
+            await TestFixOneAsync(
+    @"{ var [|x = true|];
+
+System.Diagnostics.Debug.Assert(x == true); }",
+    @"{
+        {|Warning:System.Diagnostics.Debug.Assert(true == true)|}; }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTemporary)]
+        public async Task DontWarnOnInlineIntoMultipleLocationsEvenIfConditional()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs]: 1. DoNot
2. why not?

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -200,12 +200,27 @@ private async Task<Document> InlineTemporaryAsync(Document document, VariableDec
 
             var topmostParentingExpressions = nonConflictingIdentifierNodes
                 .Select(ident => GetTopMostParentingExpression(ident))
-                .Distinct();
+                .Distinct().ToList();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]: why?

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -200,12 +200,27 @@ private async Task<Document> InlineTemporaryAsync(Document document, VariableDec
...
 
             // Make each topmost parenting statement or Equals Clause Expressions semantically explicit.
-            updatedDocument = await updatedDocument.ReplaceNodesAsync(topmostParentingExpressions, (o, n) => Simplifier.Expand(n, semanticModel, workspace, cancellationToken: cancellationToken), cancellationToken).ConfigureAwait(false);
+            updatedDocument = await updatedDocument.ReplaceNodesAsync(topmostParentingExpressions, (o, n) =>
+            {
+                var node = Simplifier.Expand(n, semanticModel, workspace, cancellationToken: cancellationToken);
+
+                // If we are inlining an expression into a single location and that single location is inside a conditional method call,
+                // then it is likely that the number of times the expression is called matters,
+                // and that the user hasn't realised that the expression won't be called unless the correct conditional symbol is defined.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]: not certain it's relevant how many times it is inlined.  i think just having the warning in each case (specifically becaues that code may not run at all) is totally reasonable.

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi

I could not think of any case where we would inline a temporary into two locations and care about one or both of them being missed of the method call is conditional.

In general if you're inlining into two locations, you don't care how often it gets called.

This means we don't have to bind the attributes in such a case, and that we reduce non-useful warnings

**Comment by jasonmalinowski** [comment]
Is this a case where we should be warning or just not showing the suggestion at all? I'm not sure I have strong opinions either way, just thought I'd ask.

I'll hold off reviewing this in detail until @dotnet/roslyn-compiler signs off on the compiler side of things that the API shape looks good. A quick glance though doesn't make me think I'll have any concerns though. :smile:

**Comment by CyrusNajmabadi** [comment]
> Is this a case where we should be warning or just not showing the suggestion at all?

We discussed this in gitter.  My take is that it should be offered.  I would just be confused if it wasn't.

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-compiler can we get a review on the compiler portion? Also which branch will that have to be in?

**Comment by jasonmalinowski** [comment]
@YairHalberstadt Are you comfortable adding compiler tests for this or need any guidance on that?

**Comment by YairHalberstadt** [comment]
@JasonMalinowski
I am happy to do so, but this will need to wait a couple of weeks, as I am on holiday at the moment.

**Comment by 333fred** [comment]
@YairHalberstadt sorry, but this is going to have to be in release/dev16.4-preview1, it's not going to meet the current compiler bar.

**Comment by YairHalberstadt** [comment]
Closed accidentally.

No problem @333Fred. This isn't exactly critical.

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]: The behavior here does seem surprising to me: if all of the locations are conditional, it still seems we'd want to warn, right? I get that sticking a warning on all of the uses might be irritating but it seems better to over-warn than be silent here.

Thoughts?

**Comment by jasonmalinowski** [comment]
@YairHalberstadt sorry for missing the compiler team had signed off on this one. IDE code looks good but I'm surprised by the design decision to not warn if it's being inlined into multiple assert locations. It seems that's still a case that we'd want to warn in...?

**Comment by YairHalberstadt** [comment]
@jasonmalinowski
My logic is that even if the locations weren't all conditional, you would have gone from calling the method once, to calling it twice, so clearly the number of times it was called doesn't matter, so it wouldn't matter if wasn't called at all.

I could definitely see that this logic is a bit tenuous though.

**Comment by jasonmalinowski** [comment]
I think I'd err  on the side of just flagging all; this doesn't pass the test I sometimes apply of "is our fancy behavior indistinguishable from a bug for a regular user?"

I still worry about the case where you have some sort of "ensure this is initialized" or "fetch some registration if it hasn't already been done" where the function in question needs to be called once, but then that's it.

**Comment by YairHalberstadt** [comment]
Sure, I can make that change.  I may have been to clever for my own good there :-)

**Comment by RikkiGibson** [comment]
@YairHalberstadt could you please merge the base branch in to resolve conflicts?

**Comment by YairHalberstadt** [comment]
Sure. Will try and do that, as well as Jason's requested change, tomorrow if I get a chance.

**Comment by YairHalberstadt** [comment]
Both changes made, @jasonmalinowski @RikkiGibson

**Comment by jasonmalinowski** [comment]
Thanks @YairHalberstadt for nice little improvement!

**Pull request merged**